Add Dropdowns to pyos website

lwasser opened the pull request:
[If you want to see a demo of the site live i forked it here](https://www.leahwasser.com/pyopensci.github.io/)

* [This is the file that defines the masthead content at the top](https://github.com/pyOpenSci/pyopensci.github.io/blob/dropdown/_includes/masthead.html)
*[ And here are the styles for that nav section (drop down specifically) ](https://github.com/pyOpenSci/pyopensci.github.io/blob/dropdown/assets/css/main.scss)

Im essentially styling over a theme. 
*****
Navigation approaches

there seems to be two approaches to these drop downs.
1. is using a button approach with links below
2. using an unordered list ul/li to create the navigation. I attempted the button approach to allow for a click drop down on mobile. 
I am not sure what approach is better in this case. 
****

There are a few issues with these right now.

1. The menu isn't lining up nicely. 

<img width="941" alt="Screen Shot 2022-10-10 at 4 15 14 PM" src="https://user-images.githubusercontent.com/7649194/194960419-22439b96-35e5-4c5c-9f8f-21fa1e218eab.png">

It also shows the hamburger too soon (which must be a width issue) but what is most weird about it is that the hamburger only works when the browser is much smaller. so i think i have different widths set somewhere in the css. 

<img width="860" alt="Screen Shot 2022-10-10 at 4 16 24 PM" src="https://user-images.githubusercontent.com/7649194/194960489-177e212c-0799-4cc4-870f-befe302f8a24.png">

finally the biggest issue  that i think is a usability thing is that i'd like the drop down to work in mobile / hamburger view when you slick on the menu. now it opens when you hover which i think is really messy.


Here is a gif that shows some of the issues
You can see me try to click on the hamburger when the screen is a bit wider and it doesn't work. so it appears too soon and doesn't show the drop down that it should when clicked. 

![website-issues-hamburger](https://user-images.githubusercontent.com/7649194/194961227-8e0337b2-2f8c-4f13-9d84-d3a2285db1f8.gif)

And here is a site that tried to create the drop down using the minimal mistakes "greedy nav" and notice the child items don't appear when you make the browser window more narrow. this is why i started from scratch.

the minimal mistakes designed users a lot of flex boxes. i am not sure if i should just create a new template with different header styles vs using the min mistakes theme. i really like the min mistakes theme a lot and like that he maintains it vs me doing that part.

darthmall reviewed (commented):
Hopefully those CSS rules will do the trick for you. I’m not sure how much the JavaScript relies on the `greedy-nav` classes from the theme. It might be a good idea to try keeping them in place if you can. It’s hard for me to know without really digging into the code of the theme, but just making these CSS changes seemed to work for me.

One other thing I thought I’d point out, if you’ll forgive more unsolicited advice, is that, personally, I might be inclined to stick with the current version of the navigation over this one. I’m not sure how much more work you were planning on the menus, but in their current form they’re not very accessible and they don’t work on mobile because they only open on hover. I tend to favor lists of links like you have currently over fancy menus because they are far more bulletproof than implementing menus in an accessible way.

If you have questions about anything or hit more road blocks, ping me here or on Twitter and I’ll take a look when I can. 😁

darthmall commented on _includes/masthead.html:
```diff
@@ -0,0 +1,131 @@
+      It's possible i am also fighting with existing styles. 
+    
+    -->
+      <nav id="site-nav" class="nav-bar-flex">
```
There are some great resources from both Mozilla and Google that might be of use, if you haven’t found them yet:

- https://developer.mozilla.org/en-US/docs/Learn
- https://web.dev/learn/

Mozilla Developer Network (MDN) has been indispensable to me my entire career.

lwasser commented:
hey @darthmall i'm making some REALLY good progress thanks to your help. 🙌  I think i understand what is generally going on.

I have a few questions:
1. right now the drop down is PUSHING all of the content below it down (so click on the hamburger and the menu pushes the entire site below down. I read here that this might have to do with how it is positioned but i'm struggling to understand how to position things correctly to make it all work properly

<img width="706" alt="Screen Shot 2022-10-11 at 2 07 14 PM" src="https://user-images.githubusercontent.com/7649194/195188846-85604f0c-d77f-4a1e-8e0a-5a6310701c1f.png">

2. Second question - in the "responsive" style the hover for drop downs still appears. I really dislike that. I'd prefer that the drop down appears on a click event. Will I have to somehow add some JS that interacts with that drop down and shows it on a click event? 

3. last question - this is a bit more minor but awkward none the less. When the screen gets smaller, the nav text bumps to two lines. I think it's rather it use flex to get smaller ( decrease font size) vs moving to two lines and then have the hamburger appear. is there a trick to do that?

Any guidance that you have on these three things is greatly appreciated. 
*****
 Note - i like your idea for the site title / logo and am not ignoring it. i just am focused on the drop down first and need to restyle the logo after to support media readers. :)

darthmall commented:
> right now the drop down is PUSHING all of the content below it down (so click on the hamburger and the menu pushes the entire site below down. I read here that this might have to do with how it is positioned but i'm struggling to understand how to position things correctly to make it all work properly

In a nutshell, you’ll want to break the menu out of normal document flow by setting `position: absolute` and then using some combination `left`, `top`, `right`, and `bottom` to position it. You’ll need to make sure that some ancestor of the menu has `position: relative` to make it the containing block for the menu (otherwise the containing block will be the document, which is not what you want).

Some relevant references:

- https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Positioning
- https://web.dev/learn/css/layout/#positioning

> Second question - in the "responsive" style the hover for drop downs still appears. I really dislike that. I'd prefer that the drop down appears on a click event. Will I have to somehow add some JS that interacts with that drop down and shows it on a click event?

Yeah, you’ll need to add some JavaScript for click events. This is where things start to get really tricky because it’s easy to do this badly and end up with something completely unusable for anyone that relies on a screen reader or keyboard.

I’ve been meaning to do more research on building accessible navigation menus just for my own edification, so if I get some time this week, I might send you a CodePen with some recommended markup and JS for building something like this that you can use as a reference. In the meantime, the Web Accessibility Initiative (WAI) has some tutorials, including one on [building menus for navigation](https://www.w3.org/WAI/tutorials/menus/). That’s where I would start for this. The [flyout menu](https://www.w3.org/WAI/tutorials/menus/flyout/) pattern is the one that’s relevant here, I think, although the sections on structure and styling are generally relevant to menus and might be worth looking through as well.

>  this is a bit more minor but awkward none the less. When the screen gets smaller, the nav text bumps to two lines. I think it's rather it use flex to get smaller ( decrease font size) vs moving to two lines and then have the hamburger appear. is there a trick to do that?

There are a few way to scale your text down, but I’m not sure I’d recommend it in this case. Text wrapping is a good thing in web design because it means your interface will remain usable on pretty much any viewport size. You never know who’s looking at your website on a smart watch or at zoom level 400%.

One approach would be to set `font-size` inside of a media query, which I believe your theme is already doing. Another approach is to use fluid typography with `clamp` (see [Utopia.fyi](https://utopia.fyi/) for an explanation and tool you can use to create fluid type scales). I personally favor fluid type scales, but making a change like that is probably going to be very invasive if your theme doesn’t support it. If that’s something you’re interested in, it’s worth considering building your own theme for Jekyll.

But I’d just let it wrap unless it’s making things illegible. Just because you’ve set the font size small enough that it doesn’t wrap _for you_ doesn’t mean it’s not wrapping for someone else if they’ve scaled up their font size, or zoomed their browser, or are using a very small device. And if you scale it down too much, you risk making it illegible to people with low vision (and then they have to zoom their browser a lot just to read this one piece of text and then zoom back out because the rest of the page has become too big).

lwasser commented:
Ok @darthmall more progress! i was able to get the drop down to work with the absolute /relative positioning switch. but now i actually have to figure out how i want the drop down to behave and also that click element on the sub menu when it's in responsive mode. 

I have a few other questions about best practices but for this menu, i've seen some examples where the drop down covers the entire page see: [ropensci.org](ropensci.org) for an example of this. i don't love how long the menu is however but it's kind of nice for that white background to appear and cover the page. Do you have any behavior suggestions for the drop down that will look good in responsive mode, media readers and in normal mode full screen? i suspect this is where it's gets super tricky as i want something that will work well across devices. i've been looking around and see all sorts of options. 

if you play with this i'm happy to look at what you figure out. ive learned a lot from those mozilla doc links that you keep sending me! [i did create a codepen](https://codepen.io/lwasser/pen/PoegxxQ) but i'm still figuring out how to port all of the styles and such over given i have so much liquid in the mix. 

<img width="785" alt="Screen Shot 2022-10-17 at 2 54 58 PM" src="https://user-images.githubusercontent.com/7649194/196282196-f667824a-76c8-47cb-89a0-996af1726166.png">
<img width="263" alt="Screen Shot 2022-10-17 at 2 55 08 PM" src="https://user-images.githubusercontent.com/7649194/196282200-deaddc03-b907-44c7-ba64-8e00a698ea3b.png">

lwasser commented:
Ok this has some really big and good updates and i'm getting closer. Struggling with a few style related position and "index" visibility elements right now.

✅ I updated the header to use a single a tag for both logo and title. it's cleaner and hopefully will be better for screen reader accessibility (still need to test this)
✅ I added hidden text in the menu for screen readers. i could also use the tag i'm not sure what is better. 
✅ I made the menu an UL as that just seems more semantically sensible / readable; I also kept the buttons - not sure if this is good or bad but i read that for accessibility buttons are easier to "target". will need to test and read more.
✅ I turned the navicon into a css element which allowed me to use css effects to create and X on open and close. 
✅ I also added a white background on the `::before` element of the responsive nav which also allowed me to let the user click anywhere to close the nav. seems to be working nicely and looks cleaner. 

So current issues that i'm struggling with:

- [ ] can't figure out how to position the "x" on the hamburger as it keeps jumping across the screen. left:0 doesn't help it. i'm wondering if there is a position up the chain that is causing that issue??
<img width="766" alt="Screen Shot 2022-10-22 at 11 33 15 AM" src="https://user-images.githubusercontent.com/7649194/197347839-cf724de5-b832-4504-bb8d-d2a49e3a5db0.png">
<img width="693" alt="Screen Shot 2022-10-22 at 11 33 09 AM" src="https://user-images.githubusercontent.com/7649194/197347841-14c6baa5-018c-4967-8451-c389519d17f3.png">

- [ ] When the navbar is shortened - things align center which makes the top look bad if text is on two lines. need a cleaner solution

<img width="676" alt="Screen Shot 2022-10-22 at 11 34 04 AM" src="https://user-images.githubusercontent.com/7649194/197347864-64444b02-8cc1-44d7-9745-020b3f9e9cc9.png">

- [ ] Hover is also behaving badly at the bottom of the nav bar element. it disappears. i suspect this might be related to zindex and size of that ul block.  but i'm not sure how to fix that yet

<img width="284" alt="Screen Shot 2022-10-22 at 11 35 39 AM" src="https://user-images.githubusercontent.com/7649194/197347936-cd0c637a-9171-4fc4-bc58-14105b61f76e.png">
<img width="254" alt="Screen Shot 2022-10-22 at 11 35 43 AM" src="https://user-images.githubusercontent.com/7649194/197347935-00362776-758e-4421-93eb-401d4c7b603e.png">

lwasser commented:
NOTE: might also add a circle ci build here to allow for in browser viewing of html files. might be nice to have for review!

lwasser commented:
Ok so things are really working just the way I want them to now. 
- [ ] The last important piece for functionality is setting up a js click listener for the drop down when in "responsive" vertical mobile etc mode. 

- [ ] And then hiding the burger until the screen is a certain width and then toggling vertical mobile friendly and horizonal nav (media queries)

- [ ] Finally do an accessibility audit using a screen reader and possibly someone who uses a screen reader too if that is possible.

darthmall commented:
Sorry for the radio silence; I let myself get pulled onto some other extracurriculars.

> I added hidden text in the menu for screen readers. i could also use the tag i'm not sure what is better

If you’re talking about line 49 or line 60 in <samp>masthead.html</samp> where you have `<span class="visually-hidden">Label</span>`, I think you’re doing the right thing. I’m not sure what “tag” your referring to as an alternative, but the way you have it ensures that the button has an accessible name (essential for interactive elements), and if CSS fails to load for some reason (it happens), the button will be labelled for sighted visitors as well.

lwasser commented:
Hey @darthmall no worries! DO you by chance have time to review this ? I need to do a bit of code cleanup but i think it may be working now.... well from what i can see it's working. i wondered if you had any other feedback?

The one thing i haven't done is gotten the "click" behavior working - but i also haven't tried. specifically in responsive mode i think i want the dropdown menus to open on click rather than hover. But i think the rest is working.

very open to your feedback here. I've also been reading a bit more about accessibility so i'm hopeful I can make this better for screen readers as well. 

Many thanks for any time in reviewing and suggesting anything you see that is flawed!  (if you have time that is)

lwasser commented:
ok HUGE progress. Nate helped me get the javascript right! 

SO NOW what's left:

- [ ] finesse the burger position - it's a bit off but works
- [ ] add aria elements for screen reader - probably will need a bit of js to toggle things on and off (burger / x)
- [ ] DO we want to have the drop down buttons have a landing page url? We might want that?  I think a user might expect that? Right now they are just button elements with no a href element.

lwasser commented:
closing this as i have a shiny new and improved one to open soon
